extensions/smallrye-health/runtime/src/main/java/io/quarkus/smallrye/health/runtime/SmallRyeHealthHandlerBase.java

```diff
@@ -41,19 +45,43 @@ private void doHandle(RoutingContext ctx) {
             Arc.container().instance(CurrentIdentityAssociation.class).get().setIdentity(user.getSecurityIdentity());
         }
         SmallRyeHealthReporter reporter = Arc.container().instance(SmallRyeHealthReporter.class).get();
-        SmallRyeHealth health = getHealth(reporter, ctx);
-        HttpServerResponse resp = ctx.response();
-        if (health.isDown()) {
-            resp.setStatusCode(503);
+        getHealth(reporter, ctx).subscribe().with(health -> {
+            HttpServerResponse resp = ctx.response();
+            if (health.isDown()) {
+                resp.setStatusCode(503);
+            }
+            resp.headers().set(HttpHeaders.CONTENT_TYPE, "application/json; charset=UTF-8");
+            try (ChunkedResponseOutputStream outputStream = new ChunkedResponseOutputStream(resp)) {
+                reporter.reportHealth(outputStream, health);
+            } catch (Throwable e) {
+                log.error("Unable to prepare health response message", e);
+            }
+            resp.end((Handler<AsyncResult<Void>>) null);
+        }, throwable -> {
+            log.error("Unable to prepare health response message", throwable);
+            HttpServerResponse resp = ctx.response();
+            resp.setStatusCode(500);
+            resp.end((Handler<AsyncResult<Void>>) null);
+        });
+    }
+
+    private static class ChunkedResponseOutputStream extends OutputStream {
+
+        private final HttpServerResponse response;
+
+        private ChunkedResponseOutputStream(HttpServerResponse response) {
+            this.response = response;
+            response.setChunked(true);
+        }
+
+        @Override
+        public void write(byte[] b, int off, int len) {
+            response.write(Buffer.buffer().appendBytes(b, off, len), null);
         }
-        resp.headers().set(HttpHeaders.CONTENT_TYPE, "application/json; charset=UTF-8");
-        Buffer buffer = Buffer.buffer(256); // this size seems to cover the basic health checks
-        try (BufferOutputStream outputStream = new BufferOutputStream(buffer);) {
-            reporter.reportHealth(outputStream, health);
-            resp.end(buffer);
-        } catch (IOException e) {
-            throw new UncheckedIOException(e);
+
+        @Override
+        public void write(int b) {
+            response.write(Buffer.buffer().appendByte((byte) b), null);
```

Comment:
Creating a buffer for every byte is super expensive. Is this method used?

Also, we need to ensure that all the writes are from the event loop, or the response will be corrupted.

Reply:
This method is implemented to satisfy the interface. The `JsonWriter` code calls only `write(byte[] b, int off, int len)` eventually. As the `write(int b)` is inefficient by design as all callers know that, this is AFAIK ok. 

As outlined in the commit history of this PR, this was introduced to avoid creating a full in-memory representation of the health status, so it is either this, or the original solution. 

It will be called from `reporter.reportHealth(outputStream, health)` so all writes are coming from this invocation without any thread switching, so I don't see a problem with the event loop here.